File: server/src/queue.ts
import { getRedis } from './cache.js'

const INGEST_QUEUE = 'xray:ingest:queue'
const INGEST_STATS = 'xray:ingest:stats'

export async function enqueueIngest(payload: Record<string, unknown>): Promise<string> {
  const r = getRedis()

  const queueItem = {
    payload,
    queued_at: new Date().toISOString(),
  }

  await r.lpush(INGEST_QUEUE, JSON.stringify(queueItem))
  await r.hincrby(INGEST_STATS, 'total_queued', 1)

  return 'queued'
}

export async function dequeueBatch(batchSize: number = 100): Promise<Record<string, unknown>[]> {
  const r = getRedis()
  const items: Record<string, unknown>[] = []

  for (let i = 0; i < batchSize; i++) {
    const item = await r.rpop(INGEST_QUEUE)
    if (!item) break

    try {
      const parsed = JSON.parse(item)
      items.push(parsed)
    } catch {
      await r.hincrby(INGEST_STATS, 'parse_errors', 1)
    }
  }

  return items
}

export async function queueLength(): Promise<number> {
  const r = getRedis()
  return r.llen(INGEST_QUEUE)
}

export async function getQueueStats(): Promise<Record<string, unknown>> {
  const r = getRedis()
  const stats = await r.hgetall(INGEST_STATS)
  const len = await queueLength()

  return {
    queue_length: len,
    total_queued: parseInt(stats.total_queued || '0'),
    total_processed: parseInt(stats.total_processed || '0'),
    total_errors: parseInt(stats.total_errors || '0'),
    parse_errors: parseInt(stats.parse_errors || '0'),
    runs_processed: parseInt(stats.runs_processed || '0'),
    steps_processed: parseInt(stats.steps_processed || '0'),
  }
}

export async function recordProcessed(runs: number = 0, steps: number = 0): Promise<void> {
  const r = getRedis()
  await r.hincrby(INGEST_STATS, 'total_processed', 1)
  if (runs > 0) await r.hincrby(INGEST_STATS, 'runs_processed', runs)
  if (steps > 0) await r.hincrby(INGEST_STATS, 'steps_processed', steps)
}

export async function recordError(): Promise<void> {
  const r = getRedis()
  await r.hincrby(INGEST_STATS, 'total_errors', 1)
}

File: server/app/queue.py
"""
Redis-backed write queue for async ingestion.

The queue buffers ingest payloads and allows background workers
to flush them to PostgreSQL/MinIO in batches, reducing write latency
for the SDK and improving throughput under load.
"""
import json
from typing import Any, Optional
from datetime import datetime

from app.cache import get_redis

# Queue keys
INGEST_QUEUE = "xray:ingest:queue"
INGEST_PROCESSING = "xray:ingest:processing"
INGEST_STATS = "xray:ingest:stats"


async def enqueue_ingest(payload: dict) -> str:
    """
    Push ingest payload to Redis queue.
    Returns a queue ID for tracking.
    """
    r = get_redis()
    
    # Add metadata
    queue_item = {
        "payload": payload,
        "queued_at": datetime.utcnow().isoformat(),
    }
    
    # Push to queue (left side - FIFO with rpop)
    await r.lpush(INGEST_QUEUE, json.dumps(queue_item, default=str))
    
    # Update stats
    await r.hincrby(INGEST_STATS, "total_queued", 1)
    
    return "queued"


async def dequeue_batch(batch_size: int = 100) -> list[dict]:
    """
    Pop a batch of payloads from the queue.
    Uses RPOP for FIFO ordering (oldest first).
    """
    r = get_redis()
    items = []
    
    for _ in range(batch_size):
        # Pop from right side (oldest items)
        item = await r.rpop(INGEST_QUEUE)
        if item is None:
            break
        
        try:
            parsed = json.loads(item)
            items.append(parsed)
        except json.JSONDecodeError:
            # Log and skip malformed items
            await r.hincrby(INGEST_STATS, "parse_errors", 1)
            continue
    
    return items


async def queue_length() -> int:
    """Get current queue length."""
    r = get_redis()
    return await r.llen(INGEST_QUEUE)


async def get_queue_stats() -> dict[str, Any]:
    """Get queue statistics."""
    r = get_redis()
    
    stats = await r.hgetall(INGEST_STATS)
    queue_len = await queue_length()
    
    return {
        "queue_length": queue_len,
        "total_queued": int(stats.get("total_queued", 0)),
        "total_processed": int(stats.get("total_processed", 0)),
        "total_errors": int(stats.get("total_errors", 0)),
        "parse_errors": int(stats.get("parse_errors", 0)),
        "runs_processed": int(stats.get("runs_processed", 0)),
        "steps_processed": int(stats.get("steps_processed", 0)),
    }


async def record_processed(runs: int = 0, steps: int = 0):
    """Record successful processing stats."""
    r = get_redis()
    await r.hincrby(INGEST_STATS, "total_processed", 1)
    if runs > 0:
        await r.hincrby(INGEST_STATS, "runs_processed", runs)
    if steps > 0:
        await r.hincrby(INGEST_STATS, "steps_processed", steps)


async def record_error():
    """Record processing error."""
    r = get_redis()
    await r.hincrby(INGEST_STATS, "total_errors", 1)


async def reset_stats():
    """Reset queue statistics (for testing)."""
    r = get_redis()
    await r.delete(INGEST_STATS)

